gadget/gadget.go

```diff
@@ -299,41 +317,99 @@ func validateVolume(name string, vol *Volume) error {
 		return fmt.Errorf("invalid schema %q", vol.Schema)
 	}
 
-	structureNames := make(map[string]bool, len(vol.Structure))
+	// named structures, for cross-referencing relative offset-write names
+	knownStructures := make(map[string]*PositionedStructure, len(vol.Structure))
+	// for validating structure overlap
+	structures := make([]PositionedStructure, len(vol.Structure))
+
+	lastOffset := Size(0)
+	farthestEnd := Size(0)
 	for idx, s := range vol.Structure {
 		if err := validateVolumeStructure(&s, vol); err != nil {
-			return fmt.Errorf("invalid structure %s: %v", fmtIndexAndName(idx, s.Name), err)
+			return fmt.Errorf("invalid structure %v: %v", fmtIndexAndName(idx, s.Name), err)
 		}
+		start := s.Offset
+		if start == 0 {
+			start = lastOffset
+		}
+		end := start + s.Size
+		ps := PositionedStructure{
+			VolumeStructure: &vol.Structure[idx],
+			StartOffset:     start,
+			index:           idx,
+		}
+		structures[idx] = ps
 		if s.Name != "" {
-			if structureNames[s.Name] {
+			if _, ok := knownStructures[s.Name]; ok {
 				return fmt.Errorf("structure name %q is not unique", s.Name)
 			}
-			structureNames[s.Name] = true
+			// keep track of named structures
+			knownStructures[s.Name] = &ps
 		}
-	}
 
-	// validate relative offsets that reference other structures by name
-	for idx, s := range vol.Structure {
-		if s.OffsetWrite.RelativeTo != "" && !structureNames[s.OffsetWrite.RelativeTo] {
-			return fmt.Errorf("structure %v refers to an unknown structure %q",
-				fmtIndexAndName(idx, s.Name), s.OffsetWrite.RelativeTo)
+		if end > farthestEnd {
+			farthestEnd = end
 		}
+		lastOffset = end
+	}
+
+	// sort by starting offset
+	sort.Sort(byStartOffset(structures))
+
+	lastEnd := Size(0)
+	// cross structure validation:
+	// - relative offsets that reference other structures by name
+	// - positioned structure overlap
+	// use structures positioned within the volume
+	for pidx, ps := range structures {
+		if ps.OffsetWrite.RelativeTo != "" {
+			// offset-write using a named structure
+			other := knownStructures[ps.OffsetWrite.RelativeTo]
+			if other == nil {
+				return fmt.Errorf("structure %v refers to an unknown structure %q",
+					fmtIndexAndName(ps.index, ps.Name), ps.OffsetWrite.RelativeTo)
```

Comment:
We could use "#index (name)" as default format for `VolumeStructure` so a call to `fmtIndexAndName()` woudn't be necessary. Or perhaps we have a recommendation to use explicit formatting? I don't know, I'm new here.

Reply:
Structure name is optional. The helper was introduced in the previous drop of validation code, specifically to handle the case when name was not provided in gadget.yaml

Reply:
I meant using the same logic of `fmtIndexAndName()` in `VolumeStructure`'s `String()` function, but that would be mostly an embellishment, and I don't know if it fits our current practices.

Reply:
Actually `PositionedStructure` has all the info to make a proper string that contains the position of the structure within the original file. I've added a Stringer support to that type.